### PR TITLE
refactor(api): refactor the implementation of windowing

### LINKED ISSUE #8847: feat(pyspark): support windowing functions in Pyspark backend
### Is your feature request related to a problem?

Windowing functions are not supported in Pyspark backend right now.

### Describe the solution you'd like

I'd like to have support for windowing functions in the Pyspark backend: https://spark.apache.org/docs/latest/structured-streaming-programming-guide.html#types-of-time-windows

### What version of ibis are you running?

main

### What backend(s) are you using, if any?

Pyspark

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by chloeh13q**
## Description of changes

Some design considerations:
- Removed the original implementation of windowing to avoid confusion.
- I thought about creating a window class to hold all the relevant windowing info like window type, window size, window slide, etc. The API would be slightly cleaner that way but I worry that it would be confused with the existing `ibis.window()`.
- Tumble and hop are the most basic types of windows that we should support across the board. Session windows and cumulate windows are not supported in every backend and the API is slightly different.

The new API:
```python
>>> import ibis
>>> from ibis import _
>>> t = ibis.table(
...     ibis.schema(
...         {
...             "createTime": "timestamp(3)",
...             "orderId": "int64",
...             "payAmount": "float64",
...             "payPlatform": "int32",
...             "provinceId": "int32",
...         }
...     ),
...     name="payment_msg",
... )
>>> expr = (
...         t.window_by(time_col="createTime")
...         .tumble(size=ibis.interval(seconds=30))
...         .agg(by=["provinceId"], avgPayAmount=_.payAmount.mean())
...     )
>>> expr
r0 := UnboundTable: payment_msg
  createTime  timestamp(3)
  orderId     int64
  payAmount   float64
  payPlatform int32
  provinceId  int32

WindowAggregate[r0]
  window_type:
    tumble
  time_col:
    r0.createTime
  groups:
    provinceId: r0.provinceId
  metrics:
    avgPayAmount: Mean(r0.payAmount)
  window_size:
    30 s
  schema:
    window_start timestamp
    window_end   timestamp
    provinceId   int32
    avgPayAmount float64
```

## Issues closed

#8847

### CODE REVIEW

**Comment by chloeh13q** [comment]
@cpcloud Thanks for reviewing!

> are there any user-facing API changes, or is it only internals that have changed?
There are user-facing API changes, although the changes are pretty subtle. We're changing:
```python
expr = (
    t.window_by(time_col="i")
    .tumble(window_size=ibis.interval(minutes=15))
    .group_by(["window_start", "window_end", "g"])
    .aggregate(mean=_.d.mean())
)
```
to something like
```python
expr = (
    t.window_by(time_col="i")
    .tumble(window_size=ibis.interval(minutes=15))
    .group_by(["g"])
    .aggregate(mean=_.d.mean())
)
```
The big difference is that we're trying to hide columns like "window_start" "window_end" from the user (as well as providing more alternative syntax in a more Ibis-like manner, e.g., you can use both `.agg` and `.aggregate`, and you can put the groupby into aggregate as well now.

**Comment by cpcloud** [comment]
@chloeh13q Can you rebase? I'll give a final review and this should be good to go after that.

**Comment by chloeh13q** [comment]
@cpcloud Just rebased!

File: ibis/backends/pyspark/tests/conftest.py [hunk 1]
```diff
@@ -409,3 +416,29 @@ def temp_table_db(con, temp_database):
...
+    if self.mode == "batch":
+        raise NotImplementedError
+    df = self._session.sql(expr.compile())
+    df.writeStream.format("memory").queryName(table_name).start()
+
+
+@pytest.fixture(autouse=True, scope="function")
+def stop_active_jobs(con_streaming):
+    yield
+    for sq in con_streaming._session.streams.active:
```
Comment (cpcloud) [inline ibis/backends/pyspark/tests/conftest.py:437]: This is a nice pattern for per-test cleanup, we might want to investigate this later for cleaning up temporary junk tables in certain cases. Not a priority, just a random thought.

File: ibis/backends/pyspark/__init__.py [hunk 2]
```diff
@@ -1109,17 +1129,34 @@ def read_parquet_dir(
         """
         path = util.normalize_filename(path)
         if self.mode == "batch":
-            spark_df = self._session.read.parquet(path, **kwargs)
+            spark_df = self._session.read
+            if schema is not None:
+                spark_df = spark_df.schema(PySparkSchema.from_ibis(schema))
+            spark_df = spark_df.parquet(path, **kwargs)
+            if watermark is not None:
+                warnings.warn("Watermark is not supported in batch mode")
```
Comment (cpcloud) [inline ibis/backends/pyspark/__init__.py:1137]: Same here, let's not give warnings that have no effect.

**Review by cpcloud — CHANGES_REQUESTED** [review]
Didn't catch the `warnings` usage before, let's clean that up and then this is good to go!

**Comment by chloeh13q** [comment]
@cpcloud Thanks for the review and feedback! I changed all the warnings to errors.

**Comment by cpcloud** [comment]
Thanks!

**Pull request merged**